network/network-address/src/encrypted.rs

```diff
@@ -0,0 +1,482 @@
+/// by restricting the visibility of their public-facing network addresses only
+/// to other validators.
+///
+/// These encrypted network addresses are intended to be stored on-chain under
+/// each validator's advertised network addresses in their `ValidatorConfig`s.
+/// All validators share the secret `root_key`, though each validator's addresses
+/// are encrypted using a per-validator derived `account_key`.
+///
+/// ### Account Key
+///
+/// ```txt
+/// account_key := HKDF-SHA3-256::extract_and_expand(
+///     salt=HKDF_SALT,
+///     ikm=root_key,
+///     info=account_address,
+///     output_length=32,
+/// )
+/// ```
+///
+/// where `hkdf_sha3_256_extract_and_expand` is
+/// [HKDF extract-and-expand](https://tools.ietf.org/html/rfc5869) with SHA3-256,
+/// `HKDF_SALT` is a constant salt for application separation, `root_key` is the
+/// shared secret distributed amongst all the validators, and `account_address`
+/// is the specific validator's [`AccountAddress`].
+///
+/// We use per-validator derived `account_key`s to limit the "blast radius" of
+/// nonce reuse to each validator, i.e., a validator that accidentally reuses a
+/// nonce will only leak information about their network addresses or derived
+/// `account_key`.
+///
+/// ### Encryption
+///
+/// A raw network address, `addr`, is then encrypted using AES-256-GCM like:
+///
+/// ```txt
+/// enc_addr := AES-256-GCM::encrypt(
+///     key=account_key,
+///     nonce=nonce,
+///     ad=root_key_version,
+///     message=addr,
+/// )
+/// ```
+///
+/// where `nonce` is a 96-bit integer as described below, `root_key_version` is
+/// the key version as a u32 big-endian integer, `addr` is the serialized
+/// `RawNetworkAddress`, and `enc_addr` is the encrypted network address
+/// concatenated with the 16-byte authentication tag.
+///
+/// ### Nonce
+///
+/// ```txt
+/// nonce := seq_num || addr_idx
+/// ```
+///
+/// where `seq_num` is the `seq_num` field as a u64 big-endian integer and
+/// `addr_idx` is the index of the encrypted network address in the list of
+/// network addresses as a u32 big-endian integer.
+///
+/// ### Sequence Number
+///
+/// In order to reduce the probability of nonce reuse, validators should use the
+/// sequence number of the rotation transaction in the `seq_num` field.
+///
+/// ### Key Rotation
+///
+/// The `EncNetworkAddress` struct contains a `root_key_version` field, which
+/// identifies the specific `root_key` used to encrypt/decrypt the
+/// `EncNetworkAddress`.
+///
+/// TODO(philiphayes): expand this section
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct EncNetworkAddress {
+    root_key_version: RootKeyVersion,
+    seq_num: u64,
+    #[serde(with = "serde_bytes")]
+    enc_addr: Vec<u8>,
+}
+
+/// Possible errors when parsing a human-readable [`NetworkAddress`].
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("error encrypting network address")]
+    EncryptError,
```

Comment:
why would an EncryptError ever happen?

Reply:
i found this weird as well. aes-gcm will fail if the plaintext length is > 2^36 bytes, but that should probably never happen

https://github.com/RustCrypto/AEADs/blob/master/aes-gcm/src/lib.rs#L234

```rust
    fn encrypt_in_place_detached(/* .. */) -> Result {
        if buffer.len() as u64 > P_MAX || associated_data.len() as u64 > A_MAX {
            return Err(Error);
        }
        // ..
    }
```

should we just unwrap here and call it a day? i also don't believe the HKDF should ever fail (since we always use valid output length. root key and account are always valid sizes too, since they're always 32 bytes), so we could unwrap that too.